R/tax_range_geo.R

```diff
@@ -0,0 +1,268 @@
+                         range_lat = rep(NA, length(unique_taxa)))
+    # Run for loop across unique taxa
+    for(i in seq_along(unique_taxa)){
+      vec <- which(occdf[, name] == unique_taxa[i])
+      lat_df$max_lat[i] <- max(occdf[vec, lat])
+      lat_df$min_lat[i] <- min(occdf[vec, lat])
+      lat_df$range_lat[i] <- lat_df$max_lat[i] - lat_df$min_lat[i]
+    }
+    # Remove row names
+    row.names(lat_df) <- NULL
+
+    # Round off values
+    lat_df[, c("max_lat", "min_lat", "range_lat")] <- round(
+      x = lat_df[, c("max_lat", "min_lat", "range_lat")], digits = 3)
+
+    # Return dataframe
+    return(lat_df)
+  }
+
+  #=== Great Circle Distance  ===
+  if (method == "gcd") {
+    # Generate dataframe for population
+    gcd_df <- data.frame()
+    # Run for loop across unique taxa
+    for(i in seq_along(unique_taxa)){
+      # Unique taxa name
+      taxa <- unique_taxa[i]
+      # taxon id
+      taxon_id <- i
+      # Subset df
+      tmp <- occdf[which(occdf[, name] == unique_taxa[i]), ]
+      # Calculate GCD matrix using the Haversine method
+      vals <- geosphere::distm(x = tmp[,c(lng, lat)],
+                               fun = geosphere::distHaversine)
+      # Convert to km
+      vals <- vals / 10^3
+      # Extract location of points with max GCD
+      loc <- which(vals == max(vals), arr.ind = TRUE)
+      # Get maximum GCD in km
+      GCD <- as.numeric(max(vals))
+      # Extract coordinates of points
+      coords <- data.frame(tmp[loc[1, 1:2], c(lng, lat)])
+      # Build dataframe
+      tmp <- cbind.data.frame(taxa, taxon_id, coords, GCD)
+      gcd_df <- rbind.data.frame(gcd_df, tmp)
+    }
+
+    # Remove row names
+    row.names(gcd_df) <- NULL
+
+    # Round off values
+    gcd_df[, c(lng, lat, "GCD")] <- round(
+      x = gcd_df[, c(lng, lat, "GCD")], digits = 3)
+
+    # Return dataframe
+    return(gcd_df)
+  }
+  #=== Occupied grid cells  ===
+  if (method == "occ") {
+    # Generate dataframe for population
+    oc_df <- data.frame(taxa = unique_taxa,
+                         taxon_id = seq(1, length(unique_taxa), 1),
+                         cells = rep(NA, length(unique_taxa)),
+                         spacing = rep(NA, length(unique_taxa)))
+    # Generate equal area hexagonal grid
+    # Which resolution should be used based on input distance/spacing?
+    # Use the h3jsr::h3_info_table to calculate resolution
+    grid <- h3jsr::h3_info_table[
+      which.min(abs(h3jsr::h3_info_table$avg_cendist_km - spacing)), ]
+    # Add resolution spacing
+    oc_df$spacing <- grid$avg_cendist_km
+    # Run for loop over all unique taxa
+    for (i in seq_along(unique_taxa)) {
+      # Subset df
+      tmp <- occdf[which(occdf[, name] == unique_taxa[i]),]
+      # Extract cell ID
+      cells <- suppressMessages(
+        h3jsr::point_to_h3(tmp[,c(lng, lat)], res = grid$h3_resolution)
+      )
+      # Calculate number of unique cells occupied
+      oc_df$cells[i] <- length(unique(cells))
```

Comment:
So the only thing I'm unsure about here is the use of what is being returned. Occupancy isn't super helpful unless there's an understanding of proportion of area occupied - by providing just which cells are occupied you obviously miss this info. Would it be possible to add an additional column to supply this info? You could use the total number of cells which all occurrences appear within as the total number, and use that to propvide a percentage of total for each unique taxa. I think this would be a more useful figure to return to the user. 

Ideally in the future, you'd want to be able to provide extent co-ordinates or some kind of mask to estimate this, but lets save that for a future edition!

Reply:
Proportional values added!